From 234576e45d8621c01c98adc870fff73ec1ed3693 Mon Sep 17 00:00:00 2001
From: Cherry Mui <cherryyz@google.com>
Date: Thu, 20 Jul 2023 14:13:51 -0400
Subject: [PATCH] cmd/link: switch to gold linker on ARM in dynamic linking mode

As reported in issue #61025, in shared linkage, when linking with
GNU linker 2.35.2 it fails. I don't have the exact version of the
GNU linker and the failure detail is unclear to me. But as we
switch to gold linker before, we can continue to do this at least
for Go 1.21.

Fixes #61025.

Change-Id: I301d9fc64e8bc82f1177c31d19b944d700cc2d7b
---

diff --git a/src/cmd/link/internal/ld/lib.go b/src/cmd/link/internal/ld/lib.go
index 91e2d51..a624ef8 100644
--- a/src/cmd/link/internal/ld/lib.go
+++ b/src/cmd/link/internal/ld/lib.go
@@ -1597,12 +1597,17 @@
 			altLinker = "lld"
 		}
 
-		if ctxt.Arch.InFamily(sys.ARM64) && buildcfg.GOOS == "linux" {
+		sharedLinkage := ctxt.BuildMode == BuildModeShared || ctxt.linkShared
+		if buildcfg.GOOS == "linux" && (ctxt.IsARM64() || (ctxt.IsARM() && sharedLinkage)) {
 			// On ARM64, the GNU linker will fail with
 			// -znocopyreloc if it thinks a COPY relocation is
 			// required. Switch to gold.
 			// https://sourceware.org/bugzilla/show_bug.cgi?id=19962
 			// https://go.dev/issue/22040
+			//
+			// On ARM, in shared linkage, linking with GNU linker 2.35.2 fails.
+			// Switch to gold.
+			// https://go.dev/issue/61025
 			altLinker = "gold"
 
 			// If gold is not installed, gcc will silently switch
@@ -1613,7 +1618,11 @@
 			cmd := exec.Command(name, args...)
 			if out, err := cmd.CombinedOutput(); err == nil {
 				if !bytes.Contains(out, []byte("GNU gold")) {
-					log.Fatalf("ARM64 external linker must be gold (issue #15696, 22040), but is not: %s", out)
+					if ctxt.IsARM64() {
+						log.Fatalf("ARM64 external linker must be gold (issue #15696, 22040), but is not: %s", out)
+					} else if ctxt.IsARM() {
+						log.Fatalf("ARM external linker must be gold (issue #61025), but is not: %s", out)
+					}
 				}
 			}
 		}
